tpu-client-next/src/tpu_client_next.rs

```diff
@@ -0,0 +1,188 @@
+//! [`ClientBuilder`] helps configuring it.
+//!
+//! # Example
+//!
+//! ```rust
+//!
+//! ```
+use {
+    crate::{
+        connection_workers_scheduler::{
+            BindTarget, ConnectionWorkersSchedulerConfig, Fanout, StakeIdentity,
+        },
+        leader_updater::LeaderUpdater,
+        transaction_batch::TransactionBatch,
+        ConnectionWorkersScheduler, SendTransactionStats,
+    },
+    solana_keypair::Keypair,
+    std::{
+        future::Future,
+        net::{SocketAddr, UdpSocket},
+        sync::Arc,
+    },
+    thiserror::Error,
+    tokio::sync::{mpsc, watch},
+    tokio_util::{sync::CancellationToken, task::TaskTracker},
+};
+
+/// [`Client`]
+#[derive(Clone)]
+pub struct Client {
+    sender: mpsc::Sender<TransactionBatch>,
+    update_certificate_sender: watch::Sender<Option<StakeIdentity>>,
+    tasks: TaskTracker,
+    cancel: CancellationToken,
+}
+
+pub struct ClientBuilder {
+    leader_updater: Box<dyn LeaderUpdater>,
+    bind_target: Option<BindTarget>,
+    identity: Option<StakeIdentity>,
+    num_connections: usize,
+    leader_send_fanout: usize,
+    skip_check_transaction_age: bool,
+    input_channel_size: usize,
+    worker_channel_size: usize,
+    max_reconnect_attempts: usize,
+    cancel: CancellationToken,
+}
+
+impl ClientBuilder {
+    pub fn with_leader_updater(leader_updater: Box<dyn LeaderUpdater>) -> Self {
+        Self {
+            leader_updater,
+            bind_target: None,
+            identity: None,
+            num_connections: 64,
+            leader_send_fanout: 2,
+            skip_check_transaction_age: true,
+            worker_channel_size: 2,
+            input_channel_size: 64,
+            max_reconnect_attempts: 2,
+            cancel: CancellationToken::new(),
+        }
+    }
+
+    pub fn bind_socket(mut self, bind_socket: UdpSocket) -> Self {
+        self.bind_target = Some(BindTarget::Socket(bind_socket));
+        self
+    }
+
+    pub fn bind_addr(mut self, bind_addr: SocketAddr) -> Self {
+        self.bind_target = Some(BindTarget::Address(bind_addr));
+        self
+    }
+
+    pub fn leader_send_fanout(mut self, fanout: usize) -> Self {
+        self.leader_send_fanout = fanout;
+        self
+    }
+
+    pub fn identity<'a>(mut self, identity: impl Into<Option<&'a Keypair>>) -> Self {
+        self.identity = identity.into().map(StakeIdentity::new);
+        self
+    }
+
+    pub fn max_cache_size(mut self, num_connections: usize) -> Self {
+        self.num_connections = num_connections;
+        self
+    }
+
+    pub fn cancel_token(mut self, cancel: CancellationToken) -> Self {
+        self.cancel = cancel;
+        self
+    }
+
+    /// TODO(klykov): API-wise, it is also possible to split the result into Sender (to
+    /// send txs) and Client which will be background task running the
+    /// scheduler. Not sure if we need this flexibility.
+    pub async fn build<F, Fut>(self, report_fn: Option<F>) -> Result<Client, ClientBuilderError>
```

Comment:
As written in the comment, this build might return two distinct objects for sending/running task. Not sure if we need such split of responsibilities. If we don't split, we have to pass Client object to the place where it is used only for sending transactions where it also can be stopped for example.